common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts

```diff
@@ -39,31 +65,90 @@ export default class LDMLKeyboardXMLSourceFileReader {
+        // retain the same subtag
+        this.boxImportsAndSpecials(sub, subtag);
       }
     } else if(typeof obj === 'object') {
       for (const key of Object.keys(obj)) {
-        if (key === 'special' || key === 'import') {
+        if (key === 'special') {
           boxXmlArray(obj, key);
+        } else if(key === 'import') {
+          // Need to 'box it up' first for processing
+          boxXmlArray(obj, key);
+          // Now, resolve the import
+          this.resolveImports(obj, subtag);
+          // now delete the import array we so carefully constructed, the caller does not
+          // want to see it.
+          delete obj['import'];
         } else {
-          this.boxImportsAndSpecials(obj[key]);
+          this.boxImportsAndSpecials(obj[key], key);
         }
       }
     }
   }
 
+  private resolveImports(obj: any, subtag: string) {
+    // first, the implied imports
+    if (subtag === 'keys') {
+      // <import base="cldr" path="techpreview/keys-Latn-implied.xml"/>
+      this.resolveOneImport(obj, subtag, {
+        base: constants.cldr_import_base,
+        path: constants.cldr_implied_keys_import
+      });
+    }
+
+    // now, the explicit imports
+    for (const asImport of (obj['import'] as LKImport[])) {
+      this.resolveOneImport(obj, subtag, asImport);
+    }
+  }
+
+  private resolveOneImport(obj: any, subtag: string, asImport: LKImport) {
+    if (asImport.base !== constants.cldr_import_base) {
+      // TODO-LDML: use custom error class CompilerImportResolveError
+      throw new Error(`import element with base ${asImport.base} is unsupported, only ${constants.cldr_import_base} is supported.`);
+    }
+    const paths = asImport.path.split('/');
+    if (paths[0] == '' || paths[1] == '' || paths.length !== 2) {
+      throw new Error(`import element with invalid path ${asImport.path}: expect the form '${constants.cldr_version_latest}./*.xml'`);
+    }
+    const importData: Uint8Array = this.readImportFile(paths[0], paths[1]);
+    if (!importData || !importData.length) {
+      throw new Error(`could not read data with path ${asImport.path}: expect the form '${constants.cldr_version_latest}./*.xml'`);
+    }
+    const importXml: any = this.loadUnboxed(importData); // TODO-LDML: have to load as any because it is an arbitrary part
+    const importRootNode = importXml[subtag]; // e.g. <keys/>
+
+    // importXml will have one property: the root element.
+    if (!importRootNode) {
+      throw new Error(`Invalid import file ${asImport.path}: expected ${subtag} as root element`);
+    }
+    // pull all children of importXml[subtag] into obj
+    for (const subsubtag of Object.keys(importRootNode)) { // e.g. <key/>
+      const subsubval = importRootNode[subsubtag];
+      if (!Array.isArray(subsubval)) {
+        throw new Error(`Problem importing ${asImport.path}: not sure how to handle non-array ${subtag}.${subsubtag}`);
+      }
+      if (!obj[subsubtag]) {
```

Comment:
I presume you are planning to move the `Error`s to `CompilerEvents` at some point?

Reply:
Yes in a follow up as noted